Speed up module/provider bump 2x

cam72cam opened the pull request:
Since the `git ls-remote` call is faster than checking RSS, it makes sense to do that first.  Modules therefore just rely on the git tags, while providers still use the RSS feed as a secondary check (to prevent additional processing for ~100 providers).

9 minutes vs 18 minutes:
https://github.com/opentofu/registry/actions/runs/7143465319

rumenvasilev commented on src/lambda/api/providerVersions.go:
```diff
@@ -96,7 +96,9 @@ func processDocument(ctx context.Context, document *providercache.VersionListing
 func fetchFromGithub(ctx context.Context, config config.Config, namespace, repoName string) (events.APIGatewayProxyResponse, error) {
```
Here we invoke the other lambda to update the DB, which essentially makes the same inventory calls as we do here with `fetchFromGithub()`: https://github.com/opentofu/registry/blob/14755e552e2c7b7955e3a50205faaa2bed1a56a0/src/lambda/api/providerVersions.go#L71-L76
A few of these repositories would easily exhaust the API limit.
I wonder if it would be a big issue if we return some kind of "retry" error back to the caller - eg, "no results found, try again in 5 mins", instead of interrogating Github's API two times for the same data. This way we'll save half the API calls and leave the inventory lambda the job of populating dynamodb.

Yantrio replied:
I think this is a valid point! I'm working on doing a writeup about an approach I'd like to take for stronger caching and to reduce the number of github api calls and I will post a comment in #97 around this.

rumenvasilev reviewed (commented):
A few suggestions here and there.

Yantrio commented on src/lambda/api/providerDownload.go:
```diff
@@ -75,3 +76,22 @@ func downloadProviderVersion(config config.Config) LambdaFunc {
+		return false, nil
+	}
+
+	return true, nil
```
This is a consuming function, and this is the error checking :)

Yantrio commented on src/internal/providers/versions.go:
```diff
@@ -25,20 +26,28 @@ type versionResult struct {
 //
 // Returns a slice of Version structures detailing each available version. If an error occurs during fetching or processing, it returns an error.
-func GetVersions(ctx context.Context, ghClient *githubv4.Client, namespace string, name string) (versions []Version, err error) {
+func GetVersions(ctx context.Context, ghClient *githubv4.Client, namespace string, name string, since *time.Time) (versions []Version, err error) {
```
I think we can add this if we determine that we need it in the long run.

Yantrio commented on src/lambda/populate_provider_versions/handler.go:
```diff
@@ -67,13 +70,23 @@ func HandleRequest(config *config.Config) LambdaFunc {
 			if err != nil {
 				return err
 			}
 
```
Note for reviewers here:

In theory this shouldn't happen, but should we also handle de-deplication of issues here, instead of just combining 2 lists of versions.

RLRabinowitz replied:
As this approach assumes that previous releases won't be deleted or updated ever, then under those assumptions it probably couldn't happen

But if, following the previous comment, we do add some buffer to the `since`, then we'd have to de-dup here
Otherwise, not too important IMO

Yantrio replied:
Do you mind if I get this merged in, then add de-duplication of versions after concatination in another PR shortly after?

tomasmik commented on src/lambda/populate_provider_versions/handler.go:
```diff
@@ -107,22 +120,29 @@ func storeVersions(ctx context.Context, e PopulateProviderVersionsEvent, version
+
+	if since == nil {
+		// check the repo exists
+		exists, err := github.RepositoryExists(ctx, config.ManagedGithubClient, e.Namespace, repoName)
```
Why even check this anyway? Only so that we can have nice errors messages? I can't see how this is handled in a different way depending on which step we fail. Also `since` can exist even if the repo doesn't anymore because it got deleted?

Yantrio replied:
As for why we check this: it's so we can short circuit and error out without consuming any of our graphql request budget. (this uses the REST api instead).

Yantrio replied:
Repo deletion is something that needs to be handled though. I need to look into how to handle that nicely. I'll raise an issue.

RLRabinowitz commented on src/lambda/populate_provider_versions/handler.go:
```diff
@@ -67,13 +70,23 @@ func HandleRequest(config *config.Config) LambdaFunc {
 					return nil
 				}
+				slog.Info("Document is stale, fetching versions", "last_updated", document.LastUpdated)
+				since = &document.LastUpdated
```
There's a technical possible race here, though extremely unlikely.
If a release was created after we fetched the GH releases, but before we saved the cache in DDB, then we'd skip it next time we populate the cache

Maybe we should add some buffer here (`&document.LastUpdated - 1 * time.minute`), though it would require us to later de-dup the versions and not just append the lists.
An alternative is to also save the `CreatedAt` of each release, and use the latest `CreatedAt` in cache instead of the `LastUpdated` of the document

Yantrio replied:
I'll add this when I add de-duplication. I'll write up an issue now

Yantrio replied:
https://github.com/opentofu/registry/issues/146

cam72cam committed:
git ls-remote is faster than checking tags

However for providers, we should also check the rss feed
for the latest entries as it's a more expensive check.

Signed-off-by: Christian Mesh <christianmesh1@gmail.com>

cam72cam committed:
Don't fail on listing tags, warn error instead

Signed-off-by: Christian Mesh <christianmesh1@gmail.com>
